File: app/src/main/java/com/mandarin/bcu/androidutil/medal/MDefiner.java
package com.mandarin.bcu.androidutil.medal;

import android.os.Environment;

import com.mandarin.bcu.androidutil.StaticStore;

import java.io.File;
import java.util.Queue;

import common.system.files.AssetData;

public class MDefiner {
    private final String FILEN = "MedalName.txt";
    private final String FILED = "MedalExplanation.txt";
    private final String[] LAN = {"/en/", "/zh/", "/kr/", "/jp/"};

    public void define() {
        if (StaticStore.medalnumber == 0) {
            String path = Environment.getExternalStorageDirectory().getPath() + "/Android/data/com.mandarin.BCU/files/org/page/medal/";

            File f = new File(path);

            if (f.exists()) {
                StaticStore.medalnumber = f.list().length;
            }
        }

        if (StaticStore.medallang == 1) {
            StaticStore.MEDNAME.clear();
            StaticStore.MEDEXP.clear();

            for (String l : LAN) {
                String path = Environment.getExternalStorageDirectory().getPath() + "/Android/data/com.mandarin.BCU/lang" + l + FILEN;

                File f = new File(path);

                if (f.exists()) {
                    Queue<String> qs = AssetData.getAsset(f).readLine();

                    if (qs != null) {
                        for (String str : qs) {
                            String[] strs = str.trim().split("\t");

                            if (strs.length == 1) {
                                continue;
                            }

                            int id = Integer.parseInt(strs[0].trim());

                            String name = strs[1].trim();

                            StaticStore.MEDNAME.put(l.substring(1, l.length() - 1), id, name);
                        }
                    }
                }

                path = Environment.getExternalStorageDirectory().getPath() + "/Android/data/com.mandarin.BCU/lang" + l + FILED;

                f = new File(path);

                if (f.exists()) {
                    Queue<String> qs = AssetData.getAsset(f).readLine();

                    if (qs != null) {
                        for (String str : qs) {
                            String[] strs = str.trim().split("\t");

                            if (strs.length == 1) {
                                continue;
                            }

                            int id = Integer.parseInt(strs[0].trim());

                            String name = strs[1].trim();

                            StaticStore.MEDEXP.put(l.substring(1, l.length() - 1), id, name);
                        }
                    }
                }
            }

            StaticStore.medallang = 0;
        }
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/medal/MDefiner.kt
package com.mandarin.bcu.androidutil.medal

import android.content.Context
import com.mandarin.bcu.androidutil.StaticStore
import common.system.files.VFile
import java.io.File

class MDefiner {
    private val filen = "MedalName.txt"
    private val filed = "MedalExplanation.txt"
    private val lan = arrayOf("/en/", "/zh/", "/kr/", "/jp/")
    fun define(c: Context) {
        if (StaticStore.medalnumber == 0) {
            val path = StaticStore.getExternalPath(c)+"org/page/medal/"

            val f = File(path)

            if (f.exists()) {
                StaticStore.medalnumber = f.list()?.size ?: 0
            }
        }

        if (StaticStore.medallang == 1) {
            StaticStore.MEDNAME.clear()
            StaticStore.MEDEXP.clear()
            for (l in lan) {
                var path = "./lang$l$filen"
                var f = File(path.replace("./", StaticStore.getExternalAsset(c)))
                if (f.exists()) {
                    val qs = VFile.readLine(path)
                    if (qs != null) {
                        for (str in qs) {
                            val strs = str.trim { it <= ' ' }.split("\t").toTypedArray()
                            if (strs.size == 1) {
                                continue
                            }
                            val id = strs[0].trim { it <= ' ' }.toInt()
                            val name = strs[1].trim { it <= ' ' }
                            StaticStore.MEDNAME.put(l.substring(1, l.length - 1), id, name)
                        }
                    }
                }

                path = "./lang$l$filed"
                f = File(path.replace("./", StaticStore.getExternalAsset(c)))

                if (f.exists()) {
                    val qs = VFile.readLine(path)
                    if (qs != null) {
                        for (str in qs) {
                            val strs = str.trim { it <= ' ' }.split("\t").toTypedArray()
                            if (strs.size == 1) {
                                continue
                            }
                            val id = strs[0].trim { it <= ' ' }.toInt()
                            val name = strs[1].trim { it <= ' ' }
                            StaticStore.MEDEXP.put(l.substring(1, l.length - 1), id, name)
                        }
                    }
                }
            }
            StaticStore.medallang = 0
        }
    }
}